### DISCUSSION TITLE
Improving TrOCR performance on LaTeX recognition tasks

### DISCUSSION THREAD

**Discussion by only-yao** [lukas-blecher/LaTeX-OCR#266] [General]
I have experimented with the trcor-base model using around 5 million LaTeX samples, but found that the performance still lags behind Mathpix. I am looking for suggestions and potential improvements to the model. Here are some optimization approaches I have already tried:

Rendering with different fonts
Adding noise to the LaTeX images
Modifying the convolutional kernel in TrOCR's image preprocessing
I have encountered some bad cases in the current model, and I would like to know if there are other strategies that could be employed to improve the model's performance on LaTeX recognition tasks.

<img width="437" alt="image" src="https://user-images.githubusercontent.com/36235579/233552671-e014c3b7-51a6-4d70-96fc-6b8888681f34.png">

<img width="554" alt="image" src="https://user-images.githubusercontent.com/36235579/233552920-6b8af3f0-a719-4c8f-9290-462909565cda.png">

Any suggestions or insights would be greatly appreciated.

**Comment by lukas-blecher** [comment]
The TrOCR resizes all images to the same size. From a quick search it looks like a square with side length 384.
That's pretty low res and doesn't really match most formulas that have a skewed aspect ratio.
I could imagine that's why there is a dot instead of a minus 
![image](https://user-images.githubusercontent.com/55287601/233632899-9653a121-9575-4a85-baaf-7fd28a067c94.png)
If you're looking for more augmentations that could help you, see pix2tex/dataset/transforms.py

https://github.com/lukas-blecher/LaTeX-OCR/blob/dd847d2ccdfbf652345dc885cbea4a1e4fe555e3/pix2tex/dataset/transforms.py#L4-L21

Also I'm going to convert this to a discussion, since it's not really an issue with this repo.

**Comment by only-yao** [comment]
hank you for your reply. Before raising the issue, I had actually made some modifications to the patch_embeddings section. The updated code is as follows:

```
class CustomViTModel(VisionEncoderDecoderModel): 
    def __init__(self, config): 
        super().__init__(config)
        self.encoder.embeddings.patch_embeddings.projection = nn.Conv2d(3, 768, kernel_size=(8, 32), stride=(8, 32))

```
With this modification, the input image size is now 192x768. However, the performance still has a significant gap compared to Mathpix.

**Comment by only-yao** [comment]
Regarding adding noise, I initially tried different noise adding methods and found that the impact on the results was not significant. Therefore, I eventually used blurring, dilation, and erosion as the noise adding methods, which are closest to real-world scenarios. The code is as follows:

```
def apply_random_noise(image, max_kernel_size=5):
    operation = random.choice(["dilate", "erode", "blur"])
    if operation == "dilate":
        kernel_size = 2
        result = apply_dilate(image, kernel_size)
    elif operation == "erode":
        kernel_size = 2
        kernel = np.ones((kernel_size, kernel_size), np.uint8)
        result = cv2.erode(image, kernel, iterations=1)
    elif operation == "blur":
        result = cv2.GaussianBlur(image, (3, 3), 0)
    return result
```
